File: main/implement-trie-prefix-tree/implement-trie-prefix-tree-scratch.py
import itertools

class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self._children = {}

    def __repr__(self):
        return 'Trie' + repr(self._children)

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        cur = self
        for ch in itertools.chain(word, ('\0',)):
            try:
                cur = cur._children[ch]
            except KeyError:
                nxt = cur._children[ch] = Trie()
                cur = nxt

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        return self.startsWith(itertools.chain(word, ('\0',)))

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        if not self._children:
            return False # so the empty trie doesn't have a "" prefix
        try:
            cur = self
            for ch in prefix:
                cur = cur._children[ch]
            return True
        except KeyError:
            return False

trie = Trie()

def test(text):
    print(f'search("{text}") -> trie.search("{trie.search(text)}")')
    print(f'startsWith("{text}") -> trie.startsWith("{trie.startsWith(text)}")')

test('')
print()

trie.insert('foobar')
trie.insert('foo')
trie.insert('quux')

for s in "", "fo", "foo", "foob", "foobar", "foobard", "quu", "quux", "quuxa":
    test(s)

print()
print(trie)

File: main/implement-trie-prefix-tree/implement-trie-prefix-tree-scratch.cpp
#include <algorithm>
#include <array>
#include <cstddef>
#include <iostream>
#include <iterator>
#include <memory>
#include <queue>
#include <string>
#include <vector>
using namespace std;

class Trie {
public:
    // Initialize your data structure here.
    Trie() = default;
    Trie(const Trie&) = delete;
    Trie(Trie&&) = default;
    Trie& operator=(Trie&) = delete;
    Trie& operator=(Trie&&) = default;
    ~Trie();

    // Inserts a word into the trie. (Basic exception guarantee only.)
    void insert(string word);

    // Checks if the word is in the trie.
    bool search(string word) const;

    // Checks if any word in the trie starts with the given prefix.
    bool startsWith(string prefix) const;

private:
    using Children = array<Trie, 'z' - 'a' + 1>;

    static size_t idx(const char ch) { return static_cast<size_t>(ch - 'a'); }

    const Trie* findNode(const string& text) const;

    unique_ptr<Children> children_ {};
    bool endsWord_ {false};
};

Trie::~Trie() {
    if (!children_) return;

    queue<unique_ptr<Children>> q;
    for (q.emplace(move(children_)); !q.empty(); q.pop()) {
        for (auto& node : *q.front())
            if (node.children_) q.emplace(move(node.children_));
    }
}

void Trie::insert(const string word) {
    auto p = this;

    for (const auto ch : word) {
        if (!p->children_) p->children_ = make_unique<Children>();
        p = &p->children_->at(idx(ch));
    }

    p->endsWord_ = true;
}

bool Trie::search(const string word) const {
    const auto p = findNode(word);
    return p != nullptr && p->endsWord_;
}

bool Trie::startsWith(const string prefix) const {
    const auto p = findNode(prefix);
    return p != nullptr && (p->children_ || p->endsWord_);
}

const Trie* Trie::findNode(const string& text) const {
    auto p = this;

    for (const auto ch : text) {
        if (!p->children_) return nullptr;
        p = &p->children_->at(idx(ch));
    }

    return p;
}

namespace {
    void run_tests() {
        Trie trie;

        const auto test = [&trie](auto text, bool show = true) {
            cout << "search(\"" << (show ? text : "...") << "\") -> \""
                 << trie.search(text) << "\"\n";
            cout << "startsWith(\"" << (show ? text : "...") << "\") -> \""
                 << trie.startsWith(text) << "\"\n";
        };

        test("");
        cout << '\n';

        trie.insert("foobar");
        trie.insert("foo");
        trie.insert("quux");

        const vector<const char*> a {
            "", "fo", "foo", "foob", "foobar", "foobard",
            "quu", "quux", "quuxa"
        };

        for_each(cbegin(a), cend(a), test);
        cout << '\n';

        string s (500, 'a');
        trie.insert(s);
        s.resize(s.size() - 1);
        test(move(s));

        string t (1'000'000, 'b');
        trie.insert(t);
        t.resize(t.size() - 1);
        test(move(t), false);

        cout << "\nDone.\n";
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cout << boolalpha;

    run_tests();
}
